core/src/main/scala/kafka/server/ReplicaAlterDirThread.scala

```diff
@@ -0,0 +1,243 @@
+  }
+
+  def buildLeaderEpochRequest(allPartitions: Seq[(TopicPartition, PartitionFetchState)]): Map[TopicPartition, Int] = {
+    val result = allPartitions
+      .filter { case (_, state) => state.isTruncatingLog }
+      .map { case (tp, _) => tp -> replicaMgr.getReplica(tp).get.epochs.get.latestEpoch }.toMap
+    result
+  }
+
+  def fetchEpochsFromLeader(partitions: Map[TopicPartition, Int]): Map[TopicPartition, EpochEndOffset] = {
+    partitions.map { case (tp, epoch) =>
+      tp -> new EpochEndOffset(Errors.NONE, replicaMgr.getReplica(tp).get.epochs.get.endOffsetFor(epoch))
+    }
+  }
+
+  def maybeTruncate(fetchedEpochs: Map[TopicPartition, EpochEndOffset]): Map[TopicPartition, Long] = {
+    val truncationPoints = scala.collection.mutable.HashMap.empty[TopicPartition, Long]
+
+    fetchedEpochs.foreach { case (topicPartition, epochOffset) =>
+      val replica = replicaMgr.getReplicaOrException(topicPartition)
+      val futureReplica = replicaMgr.getReplicaOrException(topicPartition, Request.FutureLocalReplicaId)
+
+      if (epochOffset.hasError) {
+        info(s"Retrying leaderEpoch request for partition $topicPartition as the leader reported an error: ${epochOffset.error}")
+      } else {
+        val truncationOffset =
+          if (epochOffset.endOffset() == UNDEFINED_EPOCH_OFFSET)
+            replica.highWatermark.messageOffset
+          else if (epochOffset.endOffset >= futureReplica.logEndOffset.messageOffset)
+            futureReplica.logEndOffset.messageOffset
+          else
+            epochOffset.endOffset
+        truncationPoints.put(topicPartition, truncationOffset)
+      }
+    }
+    replicaMgr.logManager.truncateTo(truncationPoints, isFuture = true)
+    truncationPoints
+  }
+
+  def buildFetchRequest(partitionMap: Seq[(TopicPartition, PartitionFetchState)]): FetchRequest = {
+    // Only include replica in the fetch request if it is not throttled.
+    val iterator = partitionMap.filter { case (topicPartition, partitionFetchState) =>
+      partitionFetchState.isReadyForFetch && !shouldThrottle(quota, topicPartition)
+    }.sortWith { (left, right) =>
+      if (left._1.topic > right._1.topic())
+        true
+      else if (left._1.topic < right._1.topic())
+        false
+      else
+        left._1.partition() > right._1.partition()
+    }.iterator
+
+    // Only move one replica at a time to increase its catch-up rate and thus reduce the time spent on moving any given replica
+    // Replicas are ordered by their TopicPartition
+    val requestMap = new util.LinkedHashMap[TopicPartition, JFetchRequest.PartitionData]
+    if (iterator.hasNext) {
+      val (topicPartition, partitionFetchState) = iterator.next()
+      val logStartOffset = replicaMgr.getReplicaOrException(topicPartition).logStartOffset
+      requestMap.put(topicPartition, new JFetchRequest.PartitionData(partitionFetchState.fetchOffset, logStartOffset, fetchSize))
```

Comment:
Hmm, the issue with this approach is that if one thread has no partition left and another thread has multiple partitions, we can't let the former to pick up the load from the latter. An alternative is to put all partitions in a queue and let AlterReplicaThreads pick up one partition at a time from the queue. This will improve the thread utilization.

Reply:
Yeah I have thought about this problem. The thing is that it requires change in the AbstractFetcherManager to address this issue, or we need to create a new manager class for ReplicaAlterDirThread. The code will be more complicated with either solution.

Currently the replica is assigned to threads using AbstractFetcherManager.getFetcherId(...). Suppose the hash function is good, it should roughly evenly balance the partition across the available ReplicaAlterDirThread. Maybe we can have a follow up JIRA to optimize this if the load imbalance across threads turn out to be an issue. Does this sound OK?